### DISCUSSION TITLE
validates decorator called after enum validation

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7184] [Usage Questions]
I have this declaration

```python
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import Enum as SaEnum
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import validates
import enum

Base = declarative_base()

class DummyEnum(enum.Enum):
    NAME_1 = "NAME_1"
    NAME_2 = "NAME_2"

class SAEnumVarcharTable(Base):
    __tablename__ = "test_table"
    id = Column(Integer, primary_key=True)
    uut_column = Column(SaEnum(DummyEnum, native_enum=False, validate_strings=True), default=DummyEnum.NAME_1)

    @validates('uut_column')
    def validate_uut_column(self, key, uut_value):
        return uut_value.upper()
```

I want the validation returning the string in upper case and only then do the Enum validation...
Is it possible?

because when I do :

```python
connection.execute(insert(SAEnumVarcharTable), {"uut_column" : "name_1"})
```
it does raise a `LookupError` where I would expect to be accepted as valid input (first goes to the validates functions -> return an uppercase string -> do the Enum validation)...Is it somehow possible?

**Comment by zzzeek** [comment]
hey there -

a validator like `@validates` only takes effect when you manipulate an instance of the object, in this case an SaEnumVarcharTable object:

```py

x = SaEnumVarcharTable()

x.uut_column = 'somename'  # <--- @validates happens here, nowhere else
```

when you invoke an insert() construct directly, your SaEnumVarcharTable class is not involved at all.

to perform validations and transformations of Python values at the Core level, like for INSERT and UPDATE statements, you would use [TypeDecorator](https://docs.sqlalchemy.org/en/14/core/custom_types.html#augmenting-existing-types).   check out the examples at https://docs.sqlalchemy.org/en/14/core/custom_types.html#typedecorator-recipes to get the idea.

**Reply by jossefaz** [reply]
Honnestly, I do not really understand this logic of parsing the enum into a key value object that we can see here :
https://github.com/sqlalchemy/sqlalchemy/blob/6aa9eeab025cecf90c1abfed430bda25aa955834/lib/sqlalchemy/sql/sqltypes.py#L1539

IMHO, this somehow breaks most of the convenient features that the enum built in lib provides.

For example, if I want to implement a `_missing_` class methods to allow and control other values to be accepted as valid, I cannot really do that. All those [useful functionality ](https://docs.python.org/3/library/enum.html#supported-sunder-names) are simply ignored by Sql Alchemy.

I would suggest to separate the ENUM type creation from its validation. The validation could be operated by the built in python enum type with all the functionalities it offers for that.

Might you accept a PR based on that idea ?

**Reply by zzzeek** [reply]
> Honnestly, I do not really understand this logic of parsing the enum into a key value object that we can see here :
> 
> https://github.com/sqlalchemy/sqlalchemy/blob/6aa9eeab025cecf90c1abfed430bda25aa955834/lib/sqlalchemy/sql/sqltypes.py#L1539
> 
> IMHO, this somehow breaks most of the convenient features that the enum built in lib provides.
> 
> For example, if I want to implement a `_missing_` class methods to allow and control other values to be accepted as valid, I cannot really do that. All those [useful functionality ](https://docs.python.org/3/library/enum.html#supported-sunder-names) are simply ignored by Sql Alchemy.

The Enum datatype in SQLAlchemy is based on [pep 435](https://www.python.org/dev/peps/pep-0435/) and was implemented before the "Enum" datatype was actually available, but rather while there were [other](https://pypi.org/project/enum/) [enum types](https://pypi.org/project/enum34/) people were using that we also wanted to support.  So there is a bigger story here that you might not have been aware of.

> 
> I would suggest to separate the ENUM type creation from its validation. The validation could be operated by the built in python enum type with all the functionalities it offers for that.

This hook appears to be introduced only in Python 3.6 so such a re-architecture would have to wait until SQLAlchemy 2.0.

also, the "Enum" type had **lots of bugs** and we dont really have the appetite to rewrite it from scratch.

A much safer approach would be to simply allow for a pluggable "missing" callable in the type that also can take effect automatically if the given pep-435 type happens to have such a method on it.

> 
> Would you accept a PR based on that idea ?

we can accept an incremental change that allows a specific hook such as "_missing_" to be useful, which must include tests and must not break any backwards incompatibility of any kind.
